src/tactic/command/global_search_trigger.py

```diff
@@ -104,7 +144,207 @@ def cleanup(my, data):
+                collection_keywords_data = keywords_data['collection']
+                for collection_code in collection_keywords_data.keys():
+                    collection = collection + " " + collection_keywords_data[collection_code]
+
+            searchable_keywords = path + " " + user + " " + collection
 
+            sobj.set_value("searchable_keywords", searchable_keywords)
+            sobj.commit(triggers=False)
+
+    def update_user_keywords(my, sobj, user_keywords):
+
+        search_key = sobj.get_search_key()
+        keywords_data = sobj.get_json_value("keywords_data", {})
+
+        keywords_data['user'] = user_keywords
+        
+        # If the collection's keywords column gets changed, all of its 
+        # children's "collection" keywords_data needs to be updated
+        if sobj.get('_is_collection'):
+            child_codes = my.get_child_codes(sobj.get_code())
+            if child_codes:
+                for child_code in child_codes:
+                    child_nest_sobject = Search.get_by_code("workflow/asset", child_code)
+                    child_nest_collection_keywords_data = child_nest_sobject.get_json_value("keywords_data", {})
+                    child_nest_collection_keywords_data['collection'][sobj.get_code()] = user_keywords
+
+                    child_nest_sobject.set_json_value("keywords_data", child_nest_collection_keywords_data)
+                    child_nest_sobject.commit(triggers=False)
+                    my.set_searchable_keywords(child_nest_sobject)
+
+        sobj.set_json_value("keywords_data", keywords_data)
+        sobj.commit(triggers=False)
+
+    def get_child_codes(my, parent_collection_code):
+
+        from pyasm.biz import Project
+        project = Project.get()
+        sql = project.get_sql()
+        database = project.get_database_type()
+        search_codes = []
+
+        if database == "SQLServer":
+            statement = '''
+            WITH res(parent_code, parent_key, search_code, search_key, path, depth) AS (
+            SELECT
+            r."parent_code", p1."name",
+            r."search_code", p2."name",
+                  CAST(r."parent_code" AS varchar(256)),
+             1
+            FROM "asset_in_asset" AS r, "asset" AS p1, "asset" AS p2
+            WHERE p1."code" IN ('%s')
+            AND p1."code" = r."parent_code" AND p2."code" = r."search_code"
+            UNION ALL
+            SELECT
+             r."parent_code", p1."name",
+             r."search_code", p2."name",
+                   CAST((path + ' > ' + r."parent_code") AS varchar(256)),
+             ng.depth + 1
+            FROM "asset_in_asset" AS r, "asset" AS p1, "asset" AS p2,
+             res AS ng
+            WHERE r."parent_code" = ng."search_code" and depth < 10
+            AND p1."code" = r."parent_code" AND p2."code" = r."search_code"
+            )
+            
+            Select search_code from res;
+            ''' % parent_collection_code
+        else:
```

Comment:
these sql statement needs to be defined in database_impl.py and called something else like get_parent_code_cte() and get_child_code_cte()